deepmd/dpmodel/descriptor/hybrid.py

```diff
@@ -0,0 +1,186 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
+from typing import (
+    List,
+    Optional,
+)
+
+import numpy as np
+
+from deepmd.dpmodel.common import (
+    NativeOP,
+)
+from deepmd.dpmodel.descriptor.base_descriptor import (
+    BaseDescriptor,
+)
+from deepmd.utils.path import (
+    DPPath,
+)
+from deepmd.utils.version import (
+    check_version_compatibility,
+)
+
+
+@BaseDescriptor.register("hybrid")
+class DescrptHybrid(BaseDescriptor, NativeOP):
+    """Concate a list of descriptors to form a new descriptor.
+
+    Parameters
+    ----------
+    list : list
+            Build a descriptor from the concatenation of the list of descriptors.
+    """
+
+    def __init__(
+        self,
+        list: list,
+    ) -> None:
+        super().__init__()
+        # warning: list is conflict with built-in list
+        descrpt_list = list
+        if descrpt_list == [] or descrpt_list is None:
+            raise RuntimeError(
+                "cannot build descriptor from an empty list of descriptors."
+            )
+        formatted_descript_list = []
+        for ii in descrpt_list:
+            if isinstance(ii, BaseDescriptor):
+                formatted_descript_list.append(ii)
+            elif isinstance(ii, dict):
+                formatted_descript_list.append(BaseDescriptor(**ii))
+            else:
+                raise NotImplementedError
+        self.descrpt_list = formatted_descript_list
+        self.numb_descrpt = len(self.descrpt_list)
+        for ii in range(1, self.numb_descrpt):
+            assert (
+                self.descrpt_list[ii].get_ntypes() == self.descrpt_list[0].get_ntypes()
+            ), f"number of atom types in {ii}th descrptor {self.descrpt_list[0].__class__.__name__} does not match others"
+        # if hybrid sel is larger than sub sel, the nlist needs to be cut for each type
+        hybrid_sel = self.get_sel()
+        self.nlist_cut_idx: List[np.ndarray] = []
+        for ii in range(self.numb_descrpt):
+            sub_sel = self.descrpt_list[ii].get_sel()
+            start_idx = np.cumsum(np.pad(hybrid_sel, (1, 0), "constant"))[:-1]
+            end_idx = start_idx + np.array(sub_sel)
+            cut_idx = np.concatenate(
+                [range(ss, ee) for ss, ee in zip(start_idx, end_idx)]
+            )
+            self.nlist_cut_idx.append(cut_idx)
+
+    def get_rcut(self) -> float:
+        """Returns the cut-off radius."""
+        return np.max([descrpt.get_rcut() for descrpt in self.descrpt_list]).item()
+
+    def get_sel(self) -> List[int]:
+        """Returns the number of selected atoms for each type."""
+        return np.max(
+            [descrpt.get_sel() for descrpt in self.descrpt_list], axis=0
```

Comment:
does it handle the case that some of the descriptors are `mixed_type` while others are not?

Reply:
It might behave differently from the theory when the `sel` is not enough. Considering a simple case, an atom has 4 type-0 neighbors and 2 type-1 neighbors, in the order of [1,0,0,0,0,1] (assume the indexes of these neighbors are 0~5). Descript 0 requests sel=[2, 3], and Descriptor 1 requests nsel=5. The expected neighbor indexes for them are `[1, 2, 0, 5, -1]` and `[0, 1, 2, 3, 4]`, respectively. When we request 5 neighbors (`nsel` of both them is 5), we cannot cover them.

If `sel` is enough, it is not a problem.

However, I noticed that the current `nlist_distinguish_types` has the same behavior... (meaning different behavior from the theory), so it might be okay.

Reply:
I agree, the distinguished type is not well supported by the pt backend.